File: AgentLens/Views/Settings/SettingsTab.swift
import SwiftUI
import OpenBurnBarCore

// MARK: - Settings Tab

/// Defines the available settings tabs in the settings navigation.
enum SettingsTab: String, CaseIterable, Identifiable {
    case home
    case general
    case aiInbox
    case updates
    case daemon
    case account
    case cloud
    case agents
    case modelProxy
    case alerts
    case notifications
    case devicesAndSync
    case textExpansion
    case media
    case dataPrivacy
    case computerUse
    case pets
    case receipts

    var id: String { rawValue }

    var title: String {
        switch self {
        case .home: return "Home"
        case .general: return "General"
        case .receipts: return "Receipts"
        case .aiInbox: return "AI Inbox"
        case .updates: return "Updates"
        case .daemon: return "Engine Room"
        case .account: return "Account"
        case .cloud: return "Cloud"
        case .agents: return "Agents"
        case .modelProxy: return "Model Proxy"
        case .alerts: return "Alerts"
        case .notifications: return "Notifications"
        case .devicesAndSync: return MacCopy.devicesAndSyncTitle
        case .textExpansion: return "Text Expansion"
        case .media: return "Media & Sharing"
        case .dataPrivacy: return "Data & Privacy"
        case .computerUse: return "Computer Use"
        case .pets: return "Pets"
        }
    }

    /// A short caption shown under the sidebar title so each entry says what
    /// lives behind it without forcing the user to click in.
    var subtitle: String {
        switch self {
        case .home:
            return "Health, attention items, and quick actions"
        case .general:
            return "Appearance, dashboard defaults, refresh, indexing, summaries"
        case .aiInbox:
            return "Background analyst, egress, cadence, and inbox budget"
        case .updates:
            return "App version, automatic updates, release channel"
        case .daemon:
            return "Daemon lifecycle, controller runtime"
        case .account:
            return "Sign-in, subscription, account actions"
        case .cloud:
            return "OpenBurnBar Cloud — hosted refresh, backup, Hermes anywhere"
        case .agents:
            return "Cloud keys, local CLIs, and local runtimes"
        case .modelProxy:
            return "Local gateway endpoint, routing strategy, model catalog"
        case .alerts:
            return "Spend thresholds, daily digest"
        case .notifications:
            return "Local pings, Telegram, calendar"
        case .devicesAndSync:
            return "Cloud sync, trusted devices, smart displays"
        case .textExpansion:
            return "&& triggers, snippets, keyboard sync, and LLM previews"
        case .media:
            return "Mercury file transfer, screen share, calls — permissions and partner preferences"
        case .dataPrivacy:
            return "Pensieve vault, exports, deletion, recovery, and panic controls"
        case .computerUse:
            return "Agent Watch, browser driving, Mac input, approvals, and audit chain"
        case .pets:
            return "Desktop companion, pet picker, agent brain, and summon hotkey"
        case .receipts:
            return "Session popups, thermal slips, quality reviews, and sounds"
        }
    }

    var icon: String {
        switch self {
        case .home: return "house.fill"
        case .general: return "gearshape.fill"
        case .receipts: return "scroll.fill"
        case .aiInbox: return "tray.full.fill"
        case .updates: return "arrow.down.circle.fill"
        case .daemon: return "cpu.fill"
        case .account: return "person.crop.circle.fill"
        case .cloud: return "sparkles"
        case .agents: return "cpu.fill"
        case .modelProxy: return "network"
        case .alerts: return "bell.fill"
        case .notifications: return "bell.badge.fill"
        case .devicesAndSync: return "macbook.and.iphone"
        case .textExpansion: return "text.cursor"
        case .media: return "play.rectangle.on.rectangle"
        case .dataPrivacy: return "lock.shield.fill"
        case .computerUse: return "cursorarrow.click.2"
        case .pets: return "pawprint.fill"
        }
    }

    /// Optional whimsical SVG asset name rendered in place of the SF Symbol
    /// for settings tabs that have a matching custom icon.
    var customIcon: String? {
        switch self {
        case .general: return "SettingsIconSettingsB"
        case .cloud: return "SettingsIconCloud"
        case .agents: return "SettingsIconAgent"
        case .account: return "SettingsIconSignIn"
        case .devicesAndSync: return "SettingsIconConnections"
        case .dataPrivacy: return "SettingsIconData"
        default: return nil
        }
    }

    var accentColor: Color {
        switch self {
        case .home: return DesignSystem.Colors.ember
        case .general: return DesignSystem.Colors.amber
        case .receipts: return DesignSystem.Colors.amber
        case .aiInbox: return DesignSystem.Colors.ember
        case .updates: return DesignSystem.Colors.frost
        case .daemon: return DesignSystem.Colors.teal
        case .account: return DesignSystem.Colors.whimsy
        case .cloud: return DesignSystem.Colors.hermesAureate
        case .agents: return DesignSystem.Colors.ember
        case .modelProxy: return DesignSystem.Colors.purple
        case .alerts: return DesignSystem.Colors.blaze
        case .notifications: return DesignSystem.Colors.whimsy
        case .devicesAndSync: return DesignSystem.Colors.teal
        case .textExpansion: return DesignSystem.Colors.amber
        case .media: return DesignSystem.Colors.hermesMercury
        case .dataPrivacy: return DesignSystem.Colors.teal
        case .computerUse: return DesignSystem.Colors.blaze
        case .pets: return DesignSystem.Colors.amber
        }
    }

    var logoProviders: [AgentProvider] {
        switch self {
        case .agents:
            return [.claudeCode, .codex, .openCode, .hermes]
        case .modelProxy:
            return [.claudeCode, .openAI, .codex]
        default:
            return []
        }
    }

    // MARK: - Section grouping

    /// Which sidebar section this tab belongs to.
    var section: SettingsSection {
        switch self {
        case .home:            return .home
        case .agents, .modelProxy:
            return .agentsAndModels
        case .general, .receipts, .aiInbox:
            return .lookAndFeel
        case .account, .cloud, .alerts, .notifications, .devicesAndSync:
            return .accountAndSync
        case .daemon, .updates, .dataPrivacy:
            return .system
        case .textExpansion, .media, .computerUse, .pets:
            return .extras
        }
    }
}

extension SettingsTab {
    /// Tabs this build ships. The build condition itself lives in
    /// `OpenBurnBarBuildGates`, which the Control Deck's `ControlKind.visibleKinds`
    /// also reads, so the sidebar and the deck can never disagree about which
    /// features exist in a given build. `ControlDeckRegistryTests` asserts they
    /// agree.
    static var visibleTabs: [SettingsTab] {
        var tabs: [SettingsTab] = [.home]
        for section in SettingsSection.visibleSections {
            tabs.append(contentsOf: section.tabs.filter(isAvailableInThisBuild))
        }
        return tabs
    }

    static func isAvailableInThisBuild(_ tab: SettingsTab) -> Bool {
        switch tab {
        case .computerUse: return OpenBurnBarBuildGates.agentControlAvailable
        case .updates: return OpenBurnBarBuildGates.updatesAvailable
        default: return true
        }
    }
}

// MARK: - Settings Section

/// Groups sidebar tabs into labeled sections so the sidebar reads as
/// organized categories rather than a flat list of entries.
enum SettingsSection: String, CaseIterable, Identifiable {
    case home
    case agentsAndModels
    case lookAndFeel
    case accountAndSync
    case system
    case extras

    var id: String { rawValue }

    var title: String {
        switch self {
        case .home:             return ""
        case .agentsAndModels:  return "Agents & Models"
        case .lookAndFeel:      return "Look & Feel"
        case .accountAndSync:   return "Account & Sync"
        case .system:           return "System"
        case .extras:           return "More"
        }
    }

    /// Tabs in display order within this section.
    var tabs: [SettingsTab] {
        switch self {
        case .home:
            return [.home]
        case .agentsAndModels:
            return [.agents, .modelProxy]
        case .lookAndFeel:
            return [.general, .receipts, .aiInbox]
        case .accountAndSync:
            return [.account, .cloud, .devicesAndSync, .alerts, .notifications]
        case .system:
            return [.daemon, .updates, .dataPrivacy]
        case .extras:
            return [.textExpansion, .media, .computerUse, .pets]
        }
    }

    /// Sections that should render in the sidebar, in order (home handled separately).
    static var visibleSections: [SettingsSection] {
        allCases.filter { $0 != .home }
    }
}

extension SettingsTab {
    /// Legacy raw values that used to identify sidebar tabs. Resolved to the
    /// new tab they were rolled into so deep links saved as
    /// `UserDefaults["settings.pendingTab"]` still land somewhere sensible.
    static func resolving(legacyRawValue raw: String) -> SettingsTab? {
        if let exact = SettingsTab(rawValue: raw) { return exact }
        switch raw {
        case "providers", "routingPools", "connections", "switcher", "hermes":
            return .agents
        case "gateway", "proxy", "modelProxy":
            return .modelProxy
        default:
            return nil
        }
    }
}

// MARK: - Shared Settings Components

struct SettingsSectionHeader: View {
    let title: String

    var body: some View {
        Text(title)
            .font(DesignSystem.Typography.caption)
            .fontWeight(.semibold)
            .foregroundStyle(DesignSystem.Colors.textSecondary)
            .padding(.top, DesignSystem.Spacing.xs)
    }
}

struct SettingsToggle: View {
    let title: String
    let subtitle: String?
    let icon: String?
    @Binding var isOn: Bool

    init(title: String, subtitle: String? = nil, icon: String? = nil, isOn: Binding<Bool>) {
        self.title = title
        self.subtitle = subtitle
        self.icon = icon
        self._isOn = isOn
    }

    var body: some View {
        Toggle(isOn: $isOn) {
            HStack(spacing: DesignSystem.Spacing.md) {
                if let icon {
                    Image(systemName: icon)
                        .font(DesignSystem.Typography.body)
                        .foregroundStyle(DesignSystem.Colors.textMuted)
                        .frame(width: 20)
                }

                VStack(alignment: .leading, spacing: 2) {
                    Text(title)
                        .font(DesignSystem.Typography.body)
                        .foregroundStyle(DesignSystem.Colors.textPrimary)

                    if let subtitle {
                        Text(subtitle)
                            .font(DesignSystem.Typography.tiny)
                            .foregroundStyle(DesignSystem.Colors.textMuted)
                    }
                }
            }
        }
        .toggleStyle(SwitchToggleStyle(tint: DesignSystem.Colors.blaze))
    }
}

File: windows/app/OpenBurnBar.App.Settings/SettingsTab.cs
// PORTED 1:1 from AgentLens/Views/Settings/SettingsTab.swift
// (enum SettingsTab + enum SettingsSection + extensions).
//
// The macOS enum also carries an accentColor (DesignSystem.Colors.*) and SF-Symbol
// icon names. Those are macOS-render concerns; this portable model keeps the parts
// the search index + shell IA need (Title, Subtitle, Section, LogoProviders, plus
// the stable RawValue + SF-symbol icon name as 1:1 data). The WinUI shell maps the
// SF-symbol name to a Segoe Fluent glyph and the tab to a Pensieve accent brush —
// see SettingsTabVisuals in windows/app/OpenBurnBar.App/Settings/.

using System.Collections.Generic;
using System.Linq;

namespace OpenBurnBar.App.Settings;

/// <summary>Available settings tabs in the sidebar navigation (Swift <c>SettingsTab</c>).</summary>
public enum SettingsTab
{
    Home,
    General,
    Updates,
    Daemon,
    Account,
    Cloud,
    Agents,
    ModelProxy,
    Alerts,
    Notifications,
    DevicesAndSync,
    TextExpansion,
    Media,
    DataPrivacy,
    ComputerUse,
    Pets,
}

/// <summary>Groups sidebar tabs into labeled sections (Swift <c>SettingsSection</c>).</summary>
public enum SettingsSection
{
    Home,
    AgentsAndModels,
    LookAndFeel,
    AccountAndSync,
    System,
    Extras,
}

/// <summary>Display + grouping metadata for <see cref="SettingsTab"/> / <see cref="SettingsSection"/>.</summary>
public static class SettingsTabMetadata
{
    /// <summary>Swift enum <c>rawValue</c> (camelCase case name) — used for legacy deep-link resolution.</summary>
    public static string RawValue(SettingsTab tab) => tab switch
    {
        SettingsTab.Home => "home",
        SettingsTab.General => "general",
        SettingsTab.Updates => "updates",
        SettingsTab.Daemon => "daemon",
        SettingsTab.Account => "account",
        SettingsTab.Cloud => "cloud",
        SettingsTab.Agents => "agents",
        SettingsTab.ModelProxy => "modelProxy",
        SettingsTab.Alerts => "alerts",
        SettingsTab.Notifications => "notifications",
        SettingsTab.DevicesAndSync => "devicesAndSync",
        SettingsTab.TextExpansion => "textExpansion",
        SettingsTab.Media => "media",
        SettingsTab.DataPrivacy => "dataPrivacy",
        SettingsTab.ComputerUse => "computerUse",
        SettingsTab.Pets => "pets",
        _ => throw new ArgumentOutOfRangeException(nameof(tab), tab, null),
    };

    public static string Title(SettingsTab tab) => tab switch
    {
        SettingsTab.Home => "Home",
        SettingsTab.General => "General",
        SettingsTab.Updates => "Updates",
        SettingsTab.Daemon => "Engine Room",
        SettingsTab.Account => "Account",
        SettingsTab.Cloud => "Cloud",
        SettingsTab.Agents => "Agents",
        SettingsTab.ModelProxy => "Model Proxy",
        SettingsTab.Alerts => "Alerts",
        SettingsTab.Notifications => "Notifications",
        SettingsTab.DevicesAndSync => "Devices & Sync",
        SettingsTab.TextExpansion => "Text Expansion",
        SettingsTab.Media => "Media & Sharing",
        SettingsTab.DataPrivacy => "Data & Privacy",
        SettingsTab.ComputerUse => "Computer Use",
        SettingsTab.Pets => "Pets",
        _ => throw new ArgumentOutOfRangeException(nameof(tab), tab, null),
    };

    public static string Subtitle(SettingsTab tab) => tab switch
    {
        SettingsTab.Home => "Health, attention items, and quick actions",
        SettingsTab.General => "Appearance, dashboard defaults, refresh, indexing, summaries",
        SettingsTab.Updates => "App version, automatic updates, release channel",
        SettingsTab.Daemon => "Daemon lifecycle, controller runtime",
        SettingsTab.Account => "Sign-in, subscription, account actions",
        SettingsTab.Cloud => "OpenBurnBar Cloud — hosted refresh, backup, Hermes anywhere",
        SettingsTab.Agents => "Cloud keys, local CLIs, and local runtimes",
        SettingsTab.ModelProxy => "Local gateway endpoint, routing strategy, model catalog",
        SettingsTab.Alerts => "Spend thresholds, daily digest",
        SettingsTab.Notifications => "Local pings, Telegram, calendar",
        SettingsTab.DevicesAndSync => "Cloud sync, trusted devices, smart displays",
        SettingsTab.TextExpansion => "&& triggers, snippets, keyboard sync, and LLM previews",
        SettingsTab.Media => "Mercury file transfer, screen share, calls — permissions and partner preferences",
        SettingsTab.DataPrivacy => "Pensieve vault, exports, deletion, recovery, and panic controls",
        SettingsTab.ComputerUse => "Agent Watch, browser driving, Mac input, approvals, and audit chain",
        SettingsTab.Pets => "Desktop companion, pet picker, agent brain, and summon hotkey",
        _ => throw new ArgumentOutOfRangeException(nameof(tab), tab, null),
    };

    /// <summary>The SF-Symbol name (macOS 1:1). The WinUI shell maps this to a Segoe Fluent glyph.</summary>
    public static string SymbolName(SettingsTab tab) => tab switch
    {
        SettingsTab.Home => "house.fill",
        SettingsTab.General => "gearshape.fill",
        SettingsTab.Updates => "arrow.down.circle.fill",
        SettingsTab.Daemon => "cpu.fill",
        SettingsTab.Account => "person.crop.circle.fill",
        SettingsTab.Cloud => "sparkles",
        SettingsTab.Agents => "cpu.fill",
        SettingsTab.ModelProxy => "network",
        SettingsTab.Alerts => "bell.fill",
        SettingsTab.Notifications => "bell.badge.fill",
        SettingsTab.DevicesAndSync => "macbook.and.iphone",
        SettingsTab.TextExpansion => "text.cursor",
        SettingsTab.Media => "play.rectangle.on.rectangle",
        SettingsTab.DataPrivacy => "lock.shield.fill",
        SettingsTab.ComputerUse => "cursorarrow.click.2",
        SettingsTab.Pets => "pawprint.fill",
        _ => throw new ArgumentOutOfRangeException(nameof(tab), tab, null),
    };

    public static IReadOnlyList<AgentProvider> LogoProviders(SettingsTab tab) => tab switch
    {
        SettingsTab.Agents => new[] { AgentProvider.ClaudeCode, AgentProvider.Codex, AgentProvider.OpenCode, AgentProvider.Hermes },
        SettingsTab.ModelProxy => new[] { AgentProvider.ClaudeCode, AgentProvider.OpenAI, AgentProvider.Codex },
        _ => Array.Empty<AgentProvider>(),
    };

    /// <summary>Which sidebar section this tab belongs to.</summary>
    public static SettingsSection Section(SettingsTab tab) => tab switch
    {
        SettingsTab.Home => SettingsSection.Home,
        SettingsTab.Agents or SettingsTab.ModelProxy => SettingsSection.AgentsAndModels,
        SettingsTab.General => SettingsSection.LookAndFeel,
        SettingsTab.Account or SettingsTab.Cloud or SettingsTab.Alerts
            or SettingsTab.Notifications or SettingsTab.DevicesAndSync => SettingsSection.AccountAndSync,
        SettingsTab.Daemon or SettingsTab.Updates or SettingsTab.DataPrivacy => SettingsSection.System,
        SettingsTab.TextExpansion or SettingsTab.Media or SettingsTab.ComputerUse or SettingsTab.Pets => SettingsSection.Extras,
        _ => throw new ArgumentOutOfRangeException(nameof(tab), tab, null),
    };

    /// <summary>All tabs in sidebar display order (home first, then each visible section).</summary>
    public static IReadOnlyList<SettingsTab> VisibleTabs
    {
        get
        {
            var tabs = new List<SettingsTab> { SettingsTab.Home };
            foreach (var section in SettingsSectionMetadata.VisibleSections)
            {
                tabs.AddRange(SettingsSectionMetadata.Tabs(section));
            }
            return tabs;
        }
    }

    /// <summary>
    /// Resolves a tab from either the current raw value or a legacy sidebar raw value
    /// (Swift <c>SettingsTab.resolving(legacyRawValue:)</c>).
    /// </summary>
    public static SettingsTab? ResolvingLegacyRawValue(string raw)
    {
        foreach (var tab in Enum.GetValues<SettingsTab>())
        {
            if (RawValue(tab) == raw) return tab;
        }
        return raw switch
        {
            "providers" or "routingPools" or "connections" or "switcher" or "hermes" => SettingsTab.Agents,
            "gateway" or "proxy" or "modelProxy" => SettingsTab.ModelProxy,
            _ => (SettingsTab?)null,
        };
    }
}

/// <summary>Section titles + ordered tab membership (Swift <c>SettingsSection</c> extension).</summary>
public static class SettingsSectionMetadata
{
    public static string Title(SettingsSection section) => section switch
    {
        SettingsSection.Home => "",
        SettingsSection.AgentsAndModels => "Agents & Models",
        SettingsSection.LookAndFeel => "Look & Feel",
        SettingsSection.AccountAndSync => "Account & Sync",
        SettingsSection.System => "System",
        SettingsSection.Extras => "More",
        _ => throw new ArgumentOutOfRangeException(nameof(section), section, null),
    };

    /// <summary>Tabs in display order within this section.</summary>
    public static IReadOnlyList<SettingsTab> Tabs(SettingsSection section) => section switch
    {
        SettingsSection.Home => new[] { SettingsTab.Home },
        SettingsSection.AgentsAndModels => new[] { SettingsTab.Agents, SettingsTab.ModelProxy },
        SettingsSection.LookAndFeel => new[] { SettingsTab.General },
        SettingsSection.AccountAndSync => new[]
        {
            SettingsTab.Account, SettingsTab.Cloud, SettingsTab.DevicesAndSync,
            SettingsTab.Alerts, SettingsTab.Notifications,
        },
        SettingsSection.System => new[] { SettingsTab.Daemon, SettingsTab.Updates, SettingsTab.DataPrivacy },
        SettingsSection.Extras => new[]
        {
            SettingsTab.TextExpansion, SettingsTab.Media, SettingsTab.ComputerUse, SettingsTab.Pets,
        },
        _ => throw new ArgumentOutOfRangeException(nameof(section), section, null),
    };

    /// <summary>Sections that render in the sidebar, in order (home handled separately).</summary>
    public static IReadOnlyList<SettingsSection> VisibleSections =>
        Enum.GetValues<SettingsSection>().Where(s => s != SettingsSection.Home).ToArray();
}
